vaadin-messages-flow-parent/vaadin-messages-flow/src/main/java/com/vaadin/flow/component/messages/MessageList.java

```diff
@@ -124,9 +128,17 @@ void scheduleItemsUpdate() {
             getElement().getNode().runWhenAttached(
                     ui -> ui.beforeClientResponse(this, ctx -> {
                         JsonArray itemsJson = JsonUtils.listToJson(items);
-                        getElement().setPropertyJson("items", itemsJson);
+                        getElement().executeJs(
+                                "window.Vaadin.Flow.messageListConnector"
+                                        + ".setItems(this, $0, $1)",
+                                itemsJson, ui.getLocale().toLanguageTag());
```

Comment:
I remembered now that this probably introduces a bug that is very common when updating a Flow component with JavaScript: If you detach and re-attach the component, the items will probably disappear.

The reason is that Flow gets rid of the client-side element instance on detach, and creates a new one on attach. Flow synchronizes all the properties from server to client, but with this change, `items` is no longer stored in the server-side `Element`'s properties.

The solution is to add an attach listener which schedules an update for the items. But it should be first tested that my assumption is correct.

We can create another ticket for that or fix in this PR, up to you.

Reply:
I tested it and items are still rendered correctly. I guess it's because of the `getElement().getNode().runWhenAttached(...)` which runs the connector when the element is attached. Do you feel we better have a test for this?

Reply:
I had to test myself and indeed it does work. :O But it's not because of `runWhenAttached`; that runs only one time, either immediately if already attached or otherwise as soon as attaching the next time.

With a breakpoint I found out that it's actually because of `LocaleChangeObserver`. For some reason `localeChange` is triggered on attach, which in turn calls `scheduleItemsUpdate`. Well, I guess it makes sense, since on attach the component gets a new reference to `UI` which owns the locale.

Anyway, now it works kind of "by accident", so I would prefer having a test case for it.

For reference, here's my changes to the test view (I guess you had something similar):
```java
addButton("detach", () -> remove(messageList));
addButton("attach", () -> addComponentAsFirst(messageList));
```

Reply:
I had exactly that 🙂